examples/seq2seq/train.py

```diff
@@ -105,72 +130,140 @@ def decode_onehot(batch_inputs):
   return np.array(list(map(decode_inputs, batch_inputs)))
 
 
+def get_sequence_lengths(sequence_batch, eos_id=1):
+  """Returns the length of each one-hot sequence, including the EOS token."""
+  eos_row = sequence_batch[:, :, eos_id]
+  eos_idx = jnp.argmax(eos_row, axis=-1)  # returns first occurence
+  return jnp.where(
+      eos_row[jnp.arange(eos_row.shape[0]), eos_idx],
+      eos_idx + 1,
+      sequence_batch.shape[1]  # if there is no EOS, use full length
+  )
+
+
+def mask_sequences(sequence_batch, lengths):
+  """Set positions beyond the length of each sequence to 0."""
+  return sequence_batch * (
+      lengths[:, np.newaxis] > np.arange(sequence_batch.shape[1]))
+
+
 class Encoder(nn.Module):
-  """LSTM encoder."""
+  """LSTM encoder, returning final state."""
 
-  def apply(self, inputs, hidden_size=512):
+  def apply(self, inputs, eos_id=1, hidden_size=512):
     # inputs.shape = (batch_size, seq_length, vocab_size).
     batch_size = inputs.shape[0]
-    carry = nn.LSTMCell.initialize_carry(jax.random.PRNGKey(0), (batch_size,),
-                                         hidden_size)
-    carry, _ = jax_utils.scan_in_dim(
-        nn.LSTMCell.partial(name='lstm'), carry, inputs, axis=1)
-    return carry
+    lengths = get_sequence_lengths(inputs, eos_id=eos_id)
+
+    lstm = nn.LSTMCell.partial(name='lstm')
+    init_lstm_state = nn.LSTMCell.initialize_carry(
+        nn.make_rng(),
+        (batch_size,),
+        hidden_size)
+
+    def encode(carry, x):
+      i, state = carry
+      new_state, y = lstm(state, x)
+      *carried_state, = jnp.where(
+          (i < lengths)[np.newaxis, :, np.newaxis], new_state, state)
+      return (i+1, tuple(carried_state)), y
+
+    (_, final_state), _ = jax_utils.scan_in_dim(
+        encode,
+        init=(0, init_lstm_state),
+        xs=inputs,
+        axis=1)
+    return final_state
 
 
 class Decoder(nn.Module):
   """LSTM decoder."""
 
-  def apply(self, carry, inputs):
+  def apply(self, init_state, inputs, teacher_force=False,
+            sample_temperature=1.0):
```

Comment:
Is it necessary to have the sample_temperature?

Right now it seems we aren't using it (it is always 1), and it also isn't directly clear to me from the code how it should be used.

Reply:
It's useful for being able to switch between "greedy" mode, sampling, etc. But, I agree that it's fine to remove since we're not using it here.